File: MergeIntervals.py
#When they hit you with the output as a List of Integer Lists
#on the Java version... it's easier to go into python...

def merge(intervals):
    if len(intervals) <= 1:
    # if we only have one or 0 intervals... just return the input as it is
        return intervals

    intervals.sort(key=lambda x:(x[0]))
    # sorting is mandatory for this algorithm, or else merging will be very messy
    # and we want to sort by the first element
    output = [intervals[0]]
    # insert the first subarray into our output
    for interval in intervals[1:]:
        if output[-1][1] >= interval[0]:
            # if the current interval we are looking at has a smaller first index
            # than the 2nd element of the previous interval
            output[-1][1] = max(output[-1][1], interval[1])
            # change the 2nd element of that previous interval
        else:
            output.append(interval)
    return output

intervals = [[1,3],[2,6],[8, 10],[15, 18]]
print(merge(intervals))

File: MergeIntervals.java
import java.util.*;

public class MergeIntervals{

    public static int[][] mergeIntervals(int[][] intervals) {
     if (intervals.length <= 1) return intervals;

        Arrays.sort(intervals, (arr1, arr2) -> Integer.compare(arr1[0], arr2[0]));
        List<int[]> output_arr = new ArrayList<>();

        int[] current = intervals[0];
        output_arr.add(current);
        
        for(int[] interval: intervals){
            int current_begin = current[0];
            int current_end = current[1];
            int next_begin = interval[0];
            int next_end = interval[1];
            if (current_end >= next_begin) current[1] = Math.max(current_end, next_end);
            else {
                current = interval;
                output_arr.add(current);
            }
            
        }
        
        return output_arr.toArray(new int[output_arr.size()][]);   
    }

    public static void main(String[] args){
        int[][] a = {{1,3}, {2,6}, {8,10}, {15,18}};
        int[][]result = mergeIntervals(a);
        System.out.println(Arrays.deepToString(result));

    }
}